### PR TITLE
Create new review endpoint & review types.

### LINKED ISSUE #12: Create review of song/album
Users should be able to post reviews of songs and albums using the same endpoint. Refer to the schema for exact field specifications.

Include validation on inputs:
* UserId and trackId/albumId must be valid
* Users cannot create multiple reviews for the same song or album (they should instead update them, which we'll implement later)
* User must include a 1-10 review, but the text comment is optional

POST (/review) a completely new review (note: return the created review object if successful)

DELIVERABLES:

1 backend endpoints, POST review
Thorough testing - include Postman screenshots of successfully created reviews and the request body needed to make them, as well as any interesting error cases

CHECK-IN: **Thursday, September 19th**, push what you have to a branch on GitHub
DEADLINE: **Monday, September 23rd**, 11:59pm, be ready to merge

### PR DESCRIPTION
**Pull request by aahiltn**
## Description

Created a working endpoint with input validations to create a new review. 
- Also created a review type and a MediaType type. 

## Issue
[Link to Ticket](https://github.com/orgs/GenerateNU/projects/35/views/1?pane=issue&itemId=79820264)

## How Has This Been Tested?
See Postman calls below. Every case should have been tested. 

## Checklist:
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] Some tests are included that test my code

## Screenshots:

Working without Comment: 
image

Working with Comment:
image

Failed by Multiple Reviews:
image

Failed by Invalid Rating: 
image

Failed by User Not Found:  -- edge case
image

Failed by Invalid MediaType: 
image

Failed by Invalid Media Id:
image

closes #12

### CODE REVIEW

File: backend/internal/supabase/migrations/20240917002108_base_schema.sql [hunk 1]
```diff
@@ -60,4 +60,6 @@ create table if not exists review (
     comment text,
     created_at timestamp with time zone default now(),
     updated_at timestamp with time zone default now()
+    -- btw if query fails due to unique constraint, id will still be incremented
+    add constraint unique_user_media unique (user_id, media_id, media_type);
```
Comment (ddusichka) [inline backend/internal/supabase/migrations/20240917002108_base_schema.sql:64]: since this base_schema migration script was already run in supabase, it'll yell at us for trying to run the same script again - instead, let's move the constraint addition to its own migration file 

to make the new migration file:
* easy one-time solution: just make a `20240921002108_add_unique_review_constraint.sql` file in `/migrations`
* proper and more helpful way is using the supabase CLI - [download it](https://supabase.com/docs/guides/cli/getting-started), cd into `backend/internal`, and try running `supabase migration new add_unique_review_constraint`. using the cli lets you spin up a local instance of the db, verify that your scripts run without errors, and see your data via a locally running dashboard, but tbh i haven't fully tested it on another machine so idk if it'll work for you guys without more setup steps. if you're feeling adventurous, you can give it a shot [(here are the docs](https://supabase.com/docs/guides/cli/local-development?queryGroups=access-method&access-method=postgres#database-migrations)) and i can help!

and this new file should just have
```
alter table review add constraint unique_user_media unique (user_id, media_id, media_type);
-- btw if query fails due to unique constraint, review.id will still be incremented
```

**Review by ddusichka — CHANGES_REQUESTED** [review]
really great work and ty for the screenshots! lmk if you have any questions about the migration

File: backend/internal/storage/postgres/schema/review.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    END AS exists
+	)
+	INSERT INTO review (user_id, media_id, media_type, rating, comment)
+	SELECT $1, $2, $3::media_type, $4, $5
+	FROM media_exists
+	WHERE exists
+	RETURNING id;
+	`
+
+	if err := r.QueryRow(ctx, query, review.UserID, review.MediaID, review.MediaType, review.Rating, review.Comment).Scan(&review.ID); err != nil {
```
Comment (garrettladley) [inline backend/internal/storage/postgres/schema/review.go:40]: some helpers for these errors would be 🔥

something like `errs.IsUniqueViolation(err error) bool` and `errs.IsForeignKeyViolation(err error, constraintName string) bool`

File: backend/internal/storage/postgres/schema/review.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		if err == pgx.ErrNoRows {
```
Comment (garrettladley) [inline backend/internal/storage/postgres/schema/review.go:41]:
```suggestion
		if errors.Is(err, pgx.ErrNoRows) {
```
prefer `if errors.Is(err, fs.ErrExist)` over `if err == fs.ErrExist`, per the [godocs](https://pkg.go.dev/errors)

**Review by ddusichka — APPROVED** [review]
looks good to me 🎉 just merge the changes from main into this

**Review by ddusichka — APPROVED** [review]
